torch/distributed/_tensor/dispatch.py

```diff
@@ -123,45 +116,68 @@ def _operator_dispatch(
+    flat_kwargs_schema: List[object] = []
+    flat_local_kwargs: List[object] = []
     mesh = None
-    for arg in arg_list:
-        if isinstance(arg, torch.Tensor) and not isinstance(arg, dtensor.DTensor):
-            raise RuntimeError(
-                f"{op_call}: got mixed torch.Tensor and DTensor, need to convert all"
-                " torch.Tensor to DTensor before calling distributed operators!"
-            )
 
+    for arg in flat_args_list:
         if isinstance(arg, dtensor.DTensor):
+            flat_args_schema.append(arg._spec)
+            flat_local_args.append(arg._local_tensor)
             if mesh is not None:
                 if mesh != arg.device_mesh:
                     raise NotImplementedError(
                         f"{op_call}: DTensor does not support cross-mesh operation yet!"
                     )
             else:
                 mesh = arg.device_mesh
+        elif isinstance(arg, torch.Tensor):
+            raise RuntimeError(
+                f"{op_call}: got mixed torch.Tensor and DTensor, need to convert all"
+                " torch.Tensor to DTensor before calling distributed operators!"
+            )
+        else:
+            flat_args_schema.append(arg)
+            flat_local_args.append(arg)
 
-    # unwrap the args/kwargs schema
-    op_schema = sharding_propagator.prepare_op_schema(op_call, args, kwargs)
-    op_schema_hash = hash(op_schema)
-
-    output_sharding = sharding_propagator.propagate(op_call, op_schema)
-
-    # first we need to lift some private aten aliases to public calls
-    if op_call in _CURRENT_DECOMPOSITION_TABLE:
-        return (
-            _CURRENT_DECOMPOSITION_TABLE[op_call](*args, **kwargs),
-            op_schema,
-            output_sharding,
-        )
-
-    # if the schema suggestion from sharding prop is not the same instance as the
-    # input op_schema, it indicates a reshard, we need to redistribute the input
-    # tensors before calling the local op
-    assert output_sharding.schema_suggestions is not None
-    suggested_input_schema = output_sharding.schema_suggestions[0]
-    needs_redistribute = hash(suggested_input_schema) != op_schema_hash
+    for kwarg in flat_kwargs_list:
+        if isinstance(kwarg, dtensor.DTensor):
+            flat_kwargs_schema.append(kwarg._spec)
+            flat_local_kwargs.append(kwarg._local_tensor)
+            if mesh is not None:
+                if mesh != kwarg.device_mesh:
+                    raise NotImplementedError(
+                        f"{op_call}: DTensor does not support cross-mesh operation yet!"
+                    )
+            else:
+                mesh = kwarg.device_mesh
+        else:
+            flat_kwargs_schema.append(kwarg)
+            flat_local_kwargs.append(kwarg)
+
+    op_info = OpInfo(
+        op_call,
+        OpSchema(
+            op_call._schema,
+            tree_unflatten(flat_args_schema, args_spec),
+            tree_unflatten(flat_kwargs_schema, kwargs_spec),
```

Comment:
Is there anyway that we can avoid this unflatten? To me this seems to be unnecessary?

Reply:
I see. Make sense, if treemap is unavoidable, ideally we only want to do it once.